rfd/0178-github-proxy.md

```diff
@@ -0,0 +1,548 @@
+aligns well with Teleport, where a Teleport user begins their day with a 'tsh'
+session, accessing only what their roleset permits. Teleport can also easily
+provide the capability to issue of short-lived client SSH certificates for
+GitHub organzations so Teleport customers do not need to implement a separate
+system for issuing these certificates. 
+
+Teleport also offers other GitHub-related features, such as [GitHub IAM
+integration](https://github.com/gravitational/teleport.e/blob/master/rfd/0021e-github-iam-integration.md)
+and GitHub SSO, where this functionality can integrate nicely. Additionally,
+proxying GitHub SSH through Teleport provides features like per-session MFA,
+audit logging.
+
+Teleport today offers a similar GitHub integration using
+[`cert_extensions`](https://github.com/gravitational/teleport/blob/branch/v16/docs/pages/admin-guides/management/guides/ssh-key-extensions.mdx)
+in the role options. This proposed GitHub proxy is considered an upgrade to the
+existing feature and should replace it.
+
+## Details
+
+### UX - User stories
+
+#### Alice configures GitHub proxy via UI
+
+Alice is a system administrator and she would like to setup the GitHub SSH CA
+integration with Teleport.
+
+Alice logs into Teleport via Web UI, and searches "github" on the
+"Enroll New Resource" page. Alice selects a "guided" GitHub integration
+experience:
+1. Alice inputs the GitHub organization "my-org", and the UI defaults (but
+   editable) to the integration name to "github-my-org".
+2. On the next page, the SSH CA and fingerprint are displayed with a link to
+   the organization's security setting page and instructions to add the CA to
+   GitHub.
+3. Next step, to setup access, Alice inputs her GitHub username.
+4. Lastly, Alice is presented instructions on how to use `tsh` to setup the Git
+   repos.
+
+#### Alice configures GitHub proxy via CLI
+
+Alice is a system administrator and she would like to setup the GitHub SSH CA
+integration with CLI.
+
+First, she creates a GitHub integration using `tctl create`:
+```yaml
+kind: integration
+sub_kind: github
+version: v1
+metadata:
+  name: github-my-org
+spec:
+  github:
+    organization: my-org
+```
+
+Then she exports the SSH CA and imports it to her GitHub organization:
+```shell
+$ tctl auth export --integration github-my-org --type github
+ssh-rsa ...
+```
+
+Next, she creates a GitHub proxy server for Teleport users to access, using
+`tctl create`:
+```yaml
+kind: node
+sub_kind: openssh-github
+version: v2
+metadata:
+  labels:
+    "github-organization": "my-org"
+spec:
+  hostname: github-my-org
+  github:
+    integration: github-my-org
+    organization: my-org
+```
+
+To provide access to this server, Alice creates a role using `tctl create`:
+```yaml
+kind: role
+metadata:
+  name: github-my-org-access
+spec:
+  allow:
+    logins:
+    - {{external.github_usernames}}
+    node_labels:
+      "github-organization": "my-org"
```

Comment:
Maybe I am overthinking this. How do we plan to differentiate between different types of Git hostings? For example, both GitHub and GitLab can have organizations. 

For the proxying purpose, what do you think something like this?
```
spec:
  allow:
    github_organizations:
    - my-org
```

Reply:
We can just call it `github_permissions` rather than generic `repo_permissions` I think. So:

```yaml
  allow:
    github_permissions:
      orgs:
      - gravitational
      repos:
      - teleport
      - teleport.e
      roles:
      - push
```